File: src/routes/graph.js
const express = require("express");
const router = express.Router();
const { $: graph } = require("../graph");

router.get("/graph", (req, res) => {
  const result = {};

  for (const key in graph) {
    const node = graph[key];

    if (node.key) {
      result[node.key] = node;

      const tmp = {
        id: node.key,
        type: node.constructor.name,
      };

      for (const prop in node) {
        const object = node[prop];

        if (typeof object === "object" && object?.key) {
          tmp[prop] = object.key;
        } else if (typeof object === "object" && !object?.key) {
          const list = [];

          for (const innerProp in object) {
            const innerObject = object[innerProp];

            if (typeof innerObject === "object" && innerObject?.key) {
              list.push(innerObject.key);
            }
          }

          if (list.length) {
            tmp[prop] = list;
          }
        }
      }

      result[key] = tmp;
    }
  }

  res.json(result);
});

module.exports = router;

File: python/src/nucleoid/routes/graph.py
"""
Graph route - provides graph visualization and inspection endpoints.
"""
from flask import Blueprint, jsonify, request
from typing import Dict, Any, List


graph_bp = Blueprint('graph', __name__)


@graph_bp.route('/', methods=['GET'])
def get_graph() -> Any:
    """
    Get the current graph state.

    Returns:
        JSON response with graph nodes and their relationships
    """
    from ..lib.graph import graph as graph_instance

    result: Dict[str, Any] = {}

    for key in graph_instance:
        node = graph_instance[key]

        if hasattr(node, 'key') and node.key:
            tmp: Dict[str, Any] = {
                'id': node.key,
                'type': node.__class__.__name__,
            }

            # Process node properties
            for prop in dir(node):
                if prop.startswith('_'):
                    continue

                try:
                    obj = getattr(node, prop)

                    # Handle single object with key
                    if hasattr(obj, 'key') and obj.key:
                        tmp[prop] = obj.key

                    # Handle dict of objects
                    elif isinstance(obj, dict) and obj:
                        obj_list: List[str] = []

                        for inner_prop in obj:
                            inner_object = obj[inner_prop]

                            if hasattr(inner_object, 'key') and inner_object.key:
                                obj_list.append(inner_object.key)

                        if obj_list:
                            tmp[prop] = obj_list

                except (AttributeError, TypeError):
                    # Skip properties that can't be accessed
                    continue

            result[key] = tmp

    return jsonify(result)
